crates/pink/pink-s3/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+#![doc = include_str!("../README.md")]
+
+use pink::chain_extension::HttpResponse;
+use pink_extension as pink;
+
+use scale::{Decode, Encode};
+// To encrypt/decrypt HTTP payloads
+use base16;
+
+// To generate AWS4 Signature
+use hmac::{Hmac, Mac};
+use sha2::Digest;
+use sha2::Sha256;
+
+#[derive(Encode, Decode, Debug, PartialEq, Eq, Copy, Clone)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum Error {
+    RequestFailed(u16),
+    InvalidEndpoint,
+}
+
+/// Infomation of a S3 object
+pub struct Head {
+    /// The size of the object
+    pub content_length: u64,
+}
+
+/// The S3 client
+pub struct S3<'a> {
+    endpoint: &'a str,
+    region: &'a str,
+    access_key: &'a str,
+    secret_key: &'a str,
+    virtual_host_mode: bool,
+}
+
+impl<'a> S3<'a> {
+    /// Create a new S3 client instance
+    pub fn new(
+        endpoint: &'a str,
+        region: &'a str,
+        access_key: &'a str,
+        secret_key: &'a str,
+    ) -> Result<Self, Error> {
+        Ok(Self {
+            endpoint,
+            region,
+            access_key,
+            secret_key,
+            virtual_host_mode: false,
+        })
+    }
+
+    /// Turn on virtual host mode
+    ///
+    /// AWS S3 requires virtual host mode for newly created buckets.
+    pub fn virtual_host_mode(self) -> Self {
+        Self {
+            virtual_host_mode: true,
+            ..self
+        }
+    }
+
+    /// Get object metadata from given bucket
+    ///
+    /// Returns Error::RequestFailed(404) it does not exist.
+    pub fn head(&self, bucket_name: &str, object_key: &str) -> Result<Head, Error> {
+        let response = self.request("HEAD", bucket_name, object_key, None)?;
+        for (k, v) in response.headers {
+            if k.to_ascii_lowercase() == "content-length" {
+                return Ok(Head {
+                    content_length: v.parse().or(Err(Error::RequestFailed(600)))?,
+                });
+            }
+        }
+        Err(Error::RequestFailed(response.status_code))
+    }
+
+    /// Get object value from bucket `bucket_name` with key `object_key`.
+    ///
+    /// Returns Error::RequestFailed(404) it does not exist.
+    pub fn get(&self, bucket_name: &str, object_key: &str) -> Result<Vec<u8>, Error> {
+        Ok(self.request("GET", bucket_name, object_key, None)?.body)
+    }
+
+    /// Put an value into bucket `bucket_name` with key `object_key`.
+    pub fn put(&self, bucket_name: &str, object_key: &str, value: &[u8]) -> Result<(), Error> {
+        self.request("PUT", bucket_name, object_key, Some(value))
+            .map(|_| ())
+    }
+
+    /// Delete given object from bucket `bucket_name` with key `object_key`.
+    ///
+    /// Returns Error::RequestFailed(404) it does not exist.
+    pub fn delete(&self, bucket_name: &str, object_key: &str) -> Result<(), Error> {
+        self.request("DELETE", bucket_name, object_key, None)
+            .map(|_| ())
+    }
+
+    fn request(
```

Comment:
Looks like better to separate the request generation and sending into two functions. This way, it can also benefit SideVM programs.

Reply:
This function also depends on the block time which is unavailable in SideVM.
I think we should port a more powerful one for SideVM that support most other S3 API features.
In fact, there is a [rusty-s3](https://github.com/paolobarbolini/rusty-s3) that exactly generates the request parameters for S3 which might be more sutable for SideVM.

Reply:
Sounds good